node/core/pvf/src/prepare/worker.rs

```diff
@@ -233,34 +303,79 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf)>
 			"prepare pvf recv_request: non utf-8 artifact path".to_string(),
 		)
 	})?;
-	Ok((code, tmp_file))
+	let preparation_timeout = framed_recv(stream).await?;
+	let preparation_timeout = Duration::decode(&mut &preparation_timeout[..]).map_err(|_| {
+		io::Error::new(
+			io::ErrorKind::Other,
+			"prepare pvf recv_request: failed to decode duration".to_string(),
+		)
+	})?;
+	Ok((code, tmp_file, preparation_timeout))
 }
 
 /// The entrypoint that the spawned prepare worker should start with. The `socket_path` specifies
 /// the path to the socket used to communicate with the host.
 pub fn worker_entrypoint(socket_path: &str) {
 	worker_event_loop("prepare", socket_path, |mut stream| async move {
 		loop {
-			let (code, dest) = recv_request(&mut stream).await?;
+			let (code, dest, preparation_timeout) = recv_request(&mut stream).await?;
 
 			gum::debug!(
 				target: LOG_TARGET,
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
 
+			// Create a lock flag. We set it when either thread finishes.
+			let lock = Arc::new(AtomicBool::new(false));
+
+			let cpu_time_start = ProcessTime::now();
+
+			// Spawn a new thread that runs the CPU time monitor. Continuously wakes up from
+			// sleeping and then either sleeps for the remaining CPU time, or kills the process if
+			// we exceed the CPU timeout.
+			let (stream_2, cpu_time_start_2, preparation_timeout_2, lock_2) =
+				(stream.clone(), cpu_time_start, preparation_timeout, lock.clone());
+			std::thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
+				task::block_on(async {
+					cpu_time_monitor_loop(
+						stream_2,
+						cpu_time_start_2,
+						preparation_timeout_2,
+						lock_2,
+					)
+					.await;
+				})
+			})?;
+
+			// Prepares the artifact in a separate thread.
 			let result = match prepare_artifact(&code) {
 				Err(err) => {
 					// Serialized error will be written into the socket.
 					Err(err)
 				},
 				Ok(compiled_artifact) => {
+					let cpu_time_elapsed = cpu_time_start.elapsed();
+
+					let result =
+						lock.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst);
```

Comment:
nit: memory orderings only matter when you have a memory access to shared resources. Since there're no accesses to reorder, all can be changed to `Relaxed`.

Reply:
Oh, is the `AtomicBool` not a shared value in this case? I mostly based the decision on this snippet:

> Relaxed operations are appropriate for things that you definitely want to happen, but don't particularly otherwise care about. For instance, incrementing a counter can be safely done by multiple threads using a relaxed fetch_add if you're not using the counter to synchronize any other accesses.

From [here](http://doc.rust-lang.org/1.65.0/nomicon/atomics.html). Since we are using the value to synchronize accesses I wanted to go with a stronger ordering. Please let me know if I'm misunderstanding.

Reply:
You're misunderstanding the purpose of memory orderings. The atomic flag itself is safe to be shared, compare-and-exchange is atomic CPU instruction and once one thread reads-modifies-writes, subsequent read will observe it.

> Memory orderings specify the way atomic operations synchronize memory. In its weakest [Ordering::Relaxed](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed), only the memory _**directly touched**_ by the operation is synchronized. On the other hand, a store-load pair of [Ordering::SeqCst](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.SeqCst) operations synchronize _**other**_ memory while additionally preserving a total order of such operations across all threads.

For reference, you can take a look at implementation of futex in Rust std, how it puts `Acquire` on the lock and `Release` on unlock (and why).

Reply:
Ah that makes sense, thanks for explaining!